### ISSUE TITLE
Add support for experimental configuration

### ISSUE THREAD

**Issue by jugglinmike** [jshint/jshint#2601]
JSHint should expose experimental/unstable functionality via a configuration that makes that status obvious to consumers. Here's what I'm proposing:

`.jshintrc` files:

```json
{
  "experimental": {
    "expnt": true,
    "async": true
  }
}
```

Inline directives:

```js
// jshint.experimental expnt: true
/* jshint.experimental async: true */
```

Although I'm using existing stage 2 proposals in the examples, I'd like to implement the configuration mechanism in advance of any specific option.

I've chosen the keyword "experimental", but I also like "unstable" since it more explicitly communicates the expectation for breaking changes between major releases. @jshint/core any thoughts on that?

**Comment by nicolo-ribaudo** [comment]
Does `experimental` imply `esversion: 6`?
If yes, does `esversion: 5` imply `experimental: false`?

**Comment by evenfrost** [comment]
Any progress in this?

**Comment by jugglinmike** [comment]
None yet. Please let us know if you'd like to help out!

**Comment by tschuy** [comment]
@jugglinmike I would be happy to help out. I've not yet touched jshint save for using it for my projects, though, so it might take me a while, but I'll start poking around this coming week. Would this be best done as one PR for both adding ``experimental: []`` and an experimental feature that can take advantage of it?

**Comment by jugglinmike** [comment]
@tschuy Great! I was hoping to implement the configuration parsing logic separately from any specific feature. It will be pretty minimal, and you'll only be able to test the option as specified via RC file (basically, just that `"experimental": {}` should not throw errors--there's no equivalent for an "empty" in-line directive).

I see that you used square brackets in your comment. Not sure if this was just a typo on your part, but solution I've described specifies this as an object value.

I'm un-assigning myself for now--looks like I was a little too optimistic when I created this issue.

**Comment by nicolo-ribaudo** [comment]
1. @tschuy Are you still working on this feature?
2. What do you think to allow `experimental.option`?
  e.g.

  ```js
// jshint option: true, experimental.option: true
  ```

  which is shorter than

  ```js
// jshint option: true
// jshint.experimental option: true
  ```

**Comment by pemasat** [comment]
Hi all,
can I  ask - are there any progress on this issue?
Thank's a lot.

**Comment by phra** [comment]
any news?

we want to start to use ES7!

**Comment by rwaldron** [comment]
@phra what feature of ES7 are you experiencing problems with when using JSHint?

**Comment by phra** [comment]
i have just tried on official website -> http://jshint.com/

try to write an ES7 async function: it takes keyword 'async' as a variable and throws error.

**Comment by rwaldron** [comment]
There is no such feature in ES7/ES2016. Here's a good blog post that lists the entirety of ES7/ES2016 features: http://www.2ality.com/2016/01/ecmascript-2016.html

**Comment by phra** [comment]
http://kangax.github.io/compat-table/esnext/

async function are at stage 3 of standardization of ES2017, they aren't already in the stage4, but babel supports them and they are **VERY** cool and time-saving.

are you planning to implement support for them?

**Comment by phra** [comment]
JSCS supports them -> http://jscs.info/rule/disallowSpacesInFunctionExpression

**Comment by rwaldron** [comment]
> async function are at stage 3 of ES2017 standardization

The feature missed ES2017 deadline. JSHint is committed to supporting stage 2 or greater, but it requires human resources and time to implement.

**Comment by phra** [comment]
do you know what files have to be changed for implement this feature?
eg: https://github.com/jshint/jshint/blob/master/src/vars.js

**Comment by rDr4g0n** [comment]
`async` and `await` are now finished and in the latest draft of the ES2017 specification. They are also now shipping in chrome and firefox. Can we reevaluate this issue?

**Comment by meteormanaged** [comment]
`async` and `await` are now live in Node 7.6.  Any update on this ?

**Comment by jugglinmike** [comment]
This issue is the definitive location for tracking the feature request, so if
there were any new information, you could expect to find it here. Would you
like to help out?

**Comment by xduseko** [comment]
Still no support for await/async? Sad...

**Comment by mikesherov** [comment]
@xduseko, JSHint is open source software maintained by an all volunteer group of contributors. 

Anyone can fork the repo, implement async/await, and submit a pull request with tests. Or, anyone can wait until the core team finds enough free time to implement it.

What were you hoping to gain with your comment?

**Comment by semmel** [comment]
@xduseko Yeah I agree it is a bit sad but eventually it will be there - I am confident. :crossed_fingers: 
If it is of any relief; right now JSLint doesn't have it either. 

Because of missing async/await I tried to migrate to ESLint but that was a great hassle. Eventually I switched back to JSHint. Any linter -- I think JSHint is still the best -- represents a tremendous value when writing JavaScript. 

To get along I am just training my eye to ignore the highlighted async/await statements in my IDE. 🙈

**Comment by johnRivs** [comment]
Unlike @semmel I can't train my eye to ignore async/await errors. In fact, if I use an undeclared variable before I use async/await, it won't show an error for the undeclared variable and that's kind of annoying.

Is there a way to make JSHint ignore async/await stuff for now?

**Comment by phra** [comment]
@johnRivs you should migrate to [eslint](http://eslint.org/) or [tslint](https://palantir.github.io/tslint/).

**Comment by dequeues** [comment]
Supporting this.

**Comment by developertenzin** [comment]
THIS. IS. AMAZING.

**Comment by mikemaccana** [comment]
Is `expnt` [exponent](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Operators/Arithmetic_Operators#Exponentiation_(**))? If so, can it please be called `exponent`? Acrns dnt hlp unstdng.

**Comment by rwaldron** [comment]
> Is expnt exponent? If so, can it please be called exponent? Acrns dnt hlp unstdng.

@jugglinmike ^^

**Comment by jugglinmike** [comment]
@mikemaccana It was intended to be, yes. The time for an "unstable" support for the exponentiation operator has come and passed, though (it will be available via `esversion: 7` in the next minor release), so we'll never see that particular option implemented. Your advice is well-heard, though, so thanks!

**Comment by patrickdark** [comment]
The most elegant solution right now seems to be a find and replace of `async` with `/* async */` and vice versa if you're validating via the JSHint website.

**Comment by jugglinmike** [comment]
@patrickdark That's a partial workaround for gh-2604, but it unfortunately doesn't account for AwaitExpressions.
